functions.gs

```diff
@@ -45,8 +61,347 @@ function checkUniqueID (uniqueid){
+//            'Prescription pickup',
+//            '26/04/2020',
+//             '',
+//            'testrequests-jb',
+//            'General supply shopping',
+//            'jb',
+//             '',
+//            'ToClose?',
+//            'https://romseymutualaid.slack.com/archives/C012HGQEJMB/p1587922239000500',
+//            '1587922239.000500',
+//             '',
+//             '',
+//            'C012HGQEJMB',
+//             'UVDT8G78T',
+//            '42',
+//            '01/05/2020 19:06:54',
+//            'New volunteer required to bring champagne to a Dr located firmly on the sofa.',
+//            ''];
+//  uniqueid='1019';
+//  userid='UVDT8G78T';
+//  channelid='C012HGQEJMB';
+  
+  
+  /// define variables
+  var globvar = globalVariables();
+  var mod_userid = globvar['MOD_USERID'];
+  var mention_mod = '<@'+mod_userid+'>';
+  
+  var log_sheetname = globvar['LOG_SHEETNAME'];
+  var tracking_sheetname = globvar['TRACKING_SHEETNAME'];
+  
+  var tracking_sheet_col_order = globvar['SHEET_COL_ORDER'];
+  var tracking_sheet_ncol = tracking_sheet_col_order.length;
+  var tracking_sheet_col_index = indexedObjectFromArray(tracking_sheet_col_order); // make associative object to easily get colindex from colname  
+  
+  var colindex_uniqueid = tracking_sheet_col_index['uniqueid'];
+  var colindex_channelid = tracking_sheet_col_index['channelid']; 
+  var colindex_phone = tracking_sheet_col_index['requesterContact'];
+  var colindex_status = tracking_sheet_col_index['requestStatus']; 
+  var colindex_volunteerName = tracking_sheet_col_index['slackVolunteerName']; 
+  var colindex_volunteerID = tracking_sheet_col_index['slackVolunteerID']; 
+  var colindex_slackts = tracking_sheet_col_index['slackTS']; 
+  var colindex_slacktsurl = tracking_sheet_col_index['slackURL'];
+  var colindex_channel = tracking_sheet_col_index['channel']; 
+  var colindex_requestername = tracking_sheet_col_index['requesterName'];
+  var colindex_address = tracking_sheet_col_index['requesterAddr'];
+  var colindex_completionCount = tracking_sheet_col_index['completionCount'];
+  var colindex_completionLastTimestamp = tracking_sheet_col_index['completionLastTimestamp']; 
+  var colindex_completionLastDetails = tracking_sheet_col_index['completionLastDetails'];  
+  
+  var channelid_true = rowvalues[colindex_channelid]; 
+  var requesterName = rowvalues[colindex_requestername];
+  var address = rowvalues[colindex_address];
+  var slackurl = rowvalues[colindex_slacktsurl];
+  var status = rowvalues[colindex_status]; 
+  var volunteerID = rowvalues[colindex_volunteerID];
+  var slack_ts = rowvalues[colindex_slackts];
+  var completionCount = rowvalues[colindex_completionCount];
+  var contactDetails = rowvalues[colindex_phone];   // private data
+  
+  var cmd_state_machine={ // this is the finite state machine object, which contains the return messages and codes for every possible {command,status} combination
```

Comment:
I like this!  

It definitely makes it easier to follow the logical flow of things, and would make it easier to explain to people.

My one concern is that the routing logic from `handleSlackCommand` is separate from this. For example on the volunteer->assigned path, there is more going on here than just a return message and this flow doesn't capture that logic. Maybe it would be better to use this exclusively for checking whether a path is valid, calling it once in `handleSlackCommand`, and leaving all state transition logic to the functions themselves?

What do you think?

Reply:
If I understand correctly you're overall suggesting to move as much redundancy in each command subfunction upwards, to `handleSlackCommand` directly? Or is there more to your comment?

To answer moving subcommand checks to `handleSlackCommand`:
- the uniqueid syntactic check can be done there
- the `checkCommandValidity` could also be done but you'd have to read the row in handleSlackCommand then?

Reply:
That's exactly that I mean. 

In an ideal world, the function for each command would focus only on implement the logic of that command only, and anything that can be done centrally should be abstracted.